cpp/src/arrow/builder.cc

```diff
@@ -1385,6 +1385,28 @@ const uint8_t* BinaryBuilder::GetValue(int64_t i, int32_t* out_length) const {
 
 StringBuilder::StringBuilder(MemoryPool* pool) : BinaryBuilder(utf8(), pool) {}
 
+Status StringBuilder::Append(const std::vector<std::string>& values,
+                             uint8_t* null_bytes) {
+  std::size_t total_length = std::accumulate(
+      values.begin(), values.end(), 0ULL,
+      [](uint64_t sum, const std::string& str) { return sum + str.size(); });
```

Comment:
Is using `std::accumulate` with a lambda the same speed as a for loop?

Reply:
Tested with below code, for-loop-solution and accumulate-solution almost has same performance.To count 100000000 strings, for-loop-solution takes 21.2312 ms,
 accumulate-solution takes 21.5120 ms.

log:
```
lizgao@lizgao-ubuntu:~/CLionProjects/cmake-test$ ./cmake_test 
WARNING: Logging before InitGoogleLogging() is written to STDERR
I0330 10:29:56.864248  3989 accmulate-lambda-performance.cc:15] Begin for_loop_solution
I0330 10:29:57.076560  3989 accmulate-lambda-performance.cc:19] End for_loop_solution = 8000000000
I0330 10:29:57.076603  3989 accmulate-lambda-performance.cc:23] Begin accumulate_solution
I0330 10:29:57.291723  3989 accmulate-lambda-performance.cc:27] End accumulate_solution = 8000000000

```

```
//
// Created by lizgao on 3/30/18.
//
#include <cstdint>
#include <numeric>
#include <string>
#include <vector>
#include <glog/logging.h>

namespace {

void for_loop_solution(const std::vector<std::string> &values) {
  std::size_t total_length = 0;
  std::size_t cnt = values.size();
  LOG(INFO) << "Begin for_loop_solution";
  for (std::size_t i = 0; i < cnt; ++i) {
    total_length += values[i].size();
  }
  LOG(INFO) << "End for_loop_solution = " << total_length;
}

void accumulate_solution(const std::vector<std::string> &values) {
  LOG(INFO) << "Begin accumulate_solution";
  std::size_t total_length = std::accumulate(
        values.begin(), values.end(), 0ULL,
        [](uint64_t sum, const std::string& str) { return sum + str.size(); });
  LOG(INFO) << "End accumulate_solution = " << total_length;
}
}

void accumulate_lambda_performance_test() {
  std::vector<std::string> str_list(100000000, std::string(80, '*'));
  for_loop_solution(str_list);
  accumulate_solution(str_list);
}
```